File: Top 100 Codes/03-Codes for Recursion/17-Find all the N bit binary numbers having more than or equal 1’s than 0’s/FindalltheNbitbinarynumbershavingmorethanorequal1sthan0s.py
def generate_binary(n, ones, zeros, current):
    """Recursive function to generate N-bit binary numbers with more 1's than 0's in all prefixes."""
    if len(current) == n:
        print(current)
        return

    # Add '1' to the current string and recurse
    generate_binary(n, ones + 1, zeros, current + "1")
    
    # Add '0' only if the number of 1's is greater than the number of 0's
    if ones > zeros:
        generate_binary(n, ones, zeros + 1, current + "0")

def print_n_bit_binary_numbers(n):
    """Function to initiate generation of N-bit binary numbers."""
    generate_binary(n, 0, 0, "")

# Take user input and call the function
n = int(input("Enter the value of N: "))
print("N-bit binary numbers having more 1's than 0's in all prefixes:")
print_n_bit_binary_numbers(n)

File: Top 100 Codes/03-Codes for Recursion/17-Find all the N bit binary numbers having more than or equal 1’s than 0’s/FindalltheNbitbinarynumbershavingmorethanorequal1sthan0s.cpp
#include <iostream>
#include <vector>
using namespace std;

// Recursive function to generate valid N-bit binary numbers
void generateBinary(int n, int ones, int zeros, string current) {
    // If the length of the current string reaches n, print it
    if (current.length() == n) {
        cout << current << endl;
        return;
    }

    // Add '1' to the current string and recurse
    generateBinary(n, ones + 1, zeros, current + "1");

    // Add '0' only if the number of 1's is greater than the number of 0's
    if (ones > zeros) {
        generateBinary(n, ones, zeros + 1, current + "0");
    }
}

// Function to print all N-bit binary numbers with more 1's than 0's in all prefixes
void printNBitBinaryNumbers(int n) {
    generateBinary(n, 0, 0, "");
}

int main() {
    int n;
    cout << "Enter the value of N: ";
    cin >> n;

    cout << "N-bit binary numbers having more 1's than 0's in all prefixes:\n";
    printNBitBinaryNumbers(n);

    return 0;
}
